src/ClassDiscovery.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Http\Discovery;
+
+/**
+ * Registry that based find results on class existence
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+abstract class ClassDiscovery
+{
+    /**
+     * Cached class
+     *
+     * @var object
+     */
+    protected static $cache;
+
+    /**
+     * Conditions and classes
+     *
+     * @var array
+     */
+    protected static $classes = [];
+
+    /**
+     * Register a factory
+     *
+     * @param string $name
+     * @param string $condition
+     * @param string $class
+     */
+    public static function register($name, $condition, $class = null)
```

Comment:
By the way, `$name` doesn’t really add much. A priority would be better. But that’s for another PR. :wink: